casper/src/main/scala/coop/rchain/casper/engine/Initializing.scala

```diff
@@ -124,12 +127,19 @@ class Initializing[F[_]
     for {
       // TODO resolve validation of approved block - we should be sure that bootstrap is not lying
       // Might be Validate.approvedBlock is enough but have to check
-      isValid <- senderIsBootstrap &&^ Validate.approvedBlock[F](approvedBlock)
+      isValid <- senderIsBootstrap &&^ Validate.approvedBlock[F](approvedBlock) &&^ shardNameIsValid.pure
```

Comment:
I believe here its better to put `shardNameIsValid.pure` before `Vaidate`, as validation is something expensive.

Reply:
Agree, shard check can be on the first place to check because the value is already computed (only wrapped in pure F).